lib/torrent.js

```diff
@@ -638,24 +639,70 @@ export default class Torrent extends EventEmitter {
     })
   }
 
-  _verifyPieces (cb) {
-    parallelLimit(this.pieces.map((piece, index) => cb => {
+  _verifyPiece (index, cb) {
+    if (this.destroyed) return cb(new Error('torrent is destroyed'))
+
+    const getOpts = {}
+    // Specify length for the last piece in case it is zero-padded
+    if (index === this.pieces.length - 1) {
+      getOpts.length = this.lastPieceLength
+    }
+    this.store.get(index, getOpts, async (err, buf) => {
       if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-      const getOpts = {}
-      // Specify length for the last piece in case it is zero-padded
-      if (index === this.pieces.length - 1) {
-        getOpts.length = this.lastPieceLength
-      }
-      this.store.get(index, getOpts, async (err, buf) => {
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+      if (err) return queueMicrotask(() => cb(null)) // ignore error
+
+      const hex = await hash(buf, 'hex')
+      if (this.destroyed) return cb(new Error('torrent is destroyed'))
 
-        if (err) return queueMicrotask(() => cb(null)) // ignore error
+      cb(null, hex === this._hashes[index])
+    })
+  }
 
-        const hex = await hash(buf, 'hex')
-        if (this.destroyed) return cb(new Error('torrent is destroyed'))
+  _verifyPiecesUsingBitfield (cb) {
+    const piecesToCheck = new Set()
+    this.files.forEach(file => {
+      let checkFile = true
+      for (let i = file._startPiece; i <= file._endPiece; ++i) {
+        if (this._startupBitfield.get(i)) {
+          if (checkFile) {
+            piecesToCheck.add(i)
+            checkFile = false
+          }
+          this._debug('piece verified %s', i)
+          this._markVerified(i)
+        } else {
+          this._markUnverified(i)
+          this._debug('piece invalid %s', i)
+        }
+      }
+    })
+    parallelLimit([...piecesToCheck].map((index) => cb => {
+      this._verifyPiece(index, (err, isVerified) => {
+        if (err) return cb(err)
+        if (!isVerified) {
+          this._startupBitfield = null
+          return cb(new Error('piece is invalid, force recheck'))
+        }
+        cb(null)
+      })
+    }), FILESYSTEM_CONCURRENCY, (err) => {
+      if (err) {
+        this._startupBitfield = null
+        return this._verifyPieces(cb)
```

Comment:
The question is "Why this file differ from what the bitfield says?", it can have two answers:
1. This specific file has been modified
2. The bitfield is not the correct one for this torrent

I think we may need more check, first of all the length of the bitfield, it should be `Math.ceil(pieces.length / 8)`
By verifying that the bitfield is the correct length the second answer is more improbable

Reply:
oh yeah, you're right, it should defo check if the bitfield length matches the torrent's bitfield size, if it doesn't don't even run any of this code

as for file being modified, this doesn't check that, realistically it only checks if it exists

Reply:
It does check that the first downloaded piece of the file matches the expected hash, it isn't perfect but the chance of two totally different files with same hash in the same piece is extremely low